### PR TITLE
Feature 2357 pulse survey report

### PR DESCRIPTION
**Pull request by mvolkmann**

### CODE REVIEW

File: server/src/main/resources/db/dev/R__Load_testing_data.sql [hunk 1]
```diff
@@ -648,8 +648,27 @@ values('dd694cf2-c0f9-4470-b897-00c564c1252b','06cd3202-a209-4ae1-a49a-10395fbe3
 INSERT INTO pulse_response
 (id, submissiondate, teammemberid, internalfeelings, externalfeelings, internal_score, external_score) -- Holly Williams
 VALUES
-('cda41eed-70ea-4d3f-a9d7-cd0c5158eb5f', '2021-01-29', '8fa673c0-ca19-4271-b759-41cb9db2e83a',  PGP_SYM_ENCRYPT('Feeling pretty happy','${aeskey}'), PGP_SYM_ENCRYPT('Feeling really good','${aeskey}'), 1, 3);
```
Comment (mvolkmann) [inline server/src/main/resources/db/dev/R__Load_testing_data.sql:651]: I added some `pulse_report` records to enable testing the new pulse report page.

File: web-ui/src/components/expand-more/ExpandMore.jsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import { IconButton } from '@mui/material';
 const ExpandMore = styled(props => {
   const { expand, ...other } = props;
   return (
-    <IconButton {...other}>
+    <IconButton classes={{ root: 'expand-more' }} {...other}>
```
Comment (mvolkmann) [inline web-ui/src/components/expand-more/ExpandMore.jsx:8]: I added a CSS class to this so I can apply CSS properties in pages were it is used.

File: web-ui/src/components/member_selector/member_selector_dialog/MemberSelectorDialog.jsx [hunk 3]
```diff
@@ -458,7 +458,6 @@ const MemberSelectorDialog = ({
           </div>
           <Button
             color="inherit"
-            disabled={checked.size === 0}
```
Comment (mvolkmann) [inline web-ui/src/components/member_selector/member_selector_dialog/MemberSelectorDialog.jsx:461]: This now allows deselecting all members.

File: web-ui/src/components/menu/Menu.jsx [hunk 4]
```diff
@@ -122,6 +123,12 @@ function Menu({ children }) {
       links.push(['/checkins-reports', 'Check-ins']);
     }
 
+    //TODO: Skipping the permission check during testing
+    //      because this permission has not been implemented yet.
+    // if (selectHasPulseReportPermission(state)) {
+    links.push(['/pulse-reports', 'Pulses']);
```
Comment (mvolkmann) [inline web-ui/src/components/menu/Menu.jsx:129]: Adds a new page to the menu.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 5]
```diff
@@ -69,7 +69,6 @@ import {
   selectCurrentUser,
   selectCurrentUserSubordinates,
   selectIsAdmin,
-  selectMyTeam,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:72]: This import was not used.

File: web-ui/src/components/reviews/periods/ReviewPeriodCard.jsx [hunk 6]
```diff
@@ -13,7 +13,6 @@ import {
   Avatar,
   Card,
   CardActions,
-  CardContent,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriodCard.jsx:16]: This import was not used.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 7]
```diff
@@ -2,7 +2,6 @@ import PropTypes from 'prop-types';
 import React, { useCallback, useContext, useEffect, useState } from 'react';
 
 import {
-  Avatar,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:5]: This import is not used.

File: web-ui/src/components/routes/Routes.jsx [hunk 8]
```diff
@@ -147,6 +148,10 @@ export default function Routes() {
             <Header title="Annual Review Report" />
             <AnnualReviewReportPage />
           </Route>
+          <Route path="/pulse-reports">
```
Comment (mvolkmann) [inline web-ui/src/components/routes/Routes.jsx:151]: New route for new page.

File: web-ui/src/context/selectors.js [hunk 9]
```diff
@@ -94,6 +94,16 @@ export const selectHasProfileReportPermission = createSelector(
     )
 );
 
+export const selectHasPulseReportPermission = createSelector(
```
Comment (mvolkmann) [inline web-ui/src/context/selectors.js:97]: New permission check.

File: web-ui/src/pages/PulsePage.jsx [hunk 10]
```diff
@@ -15,9 +15,9 @@ const PulsePage = () => {
   const csrf = selectCsrfToken(state);
 
   const [externalComment, setExternalComment] = useState('');
-  const [externalScore, setExternalScore] = useState(0);
```
Comment (mvolkmann) [inline web-ui/src/pages/PulsePage.jsx:18]: Updated default internal and external scores.
These are 0-based in the code (not in the database), so 2 is in the middle.

File: web-ui/src/pages/PulsePage.jsx [hunk 11]
```diff
@@ -79,13 +79,11 @@ const PulsePage = () => {
 
   const submit = async () => {
     const myId = currentUser?.id;
-    //TODO: The POST endpoint doesn't currently save the score values,
```
Comment (mvolkmann) [inline web-ui/src/pages/PulsePage.jsx:82]: This TODO has been addressed.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 12]
```diff
@@ -0,0 +1,402 @@
+import { format } from 'date-fns';
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:1]: This is the new report page.

File: web-ui/src/pages/PulsePage.jsx [hunk 13]
```diff
@@ -102,6 +100,9 @@ const PulsePage = () => {
         data
       });
       if (res.error) throw new Error(res.error.message);
+
+      // Refresh browser to show that pulses where already submitted today.
+      location.reload();
```
Comment (vhscom) [inline web-ui/src/pages/PulsePage.jsx:105]: I noticed the page reloads nicely using `react-router-dom`'s `useHistory` hook like:

```js
import {useHistory} from 'react-router-dom';

const PulsePage = () => {
  const { state } = useContext(AppContext);
  const currentUser = selectCurrentUser(state);
  const csrf = selectCsrfToken(state);
  const history = useHistory();
```

Then the hook can be used to navigate the same page:

```js
history.push(location.pathname);
```

The user will then see a gentle page update and the thank you message.

**Review by vhscom — APPROVED** [review]
Looking great! One change suggested for consideration.

**Review by S78901 — APPROVED** [review]
Looks good, agree with @vhscom on the router update

File: web-ui/src/pages/PulsePage.jsx (hunk 13 above)
Comment (mvolkmann) [reply web-ui/src/pages/PulsePage.jsx]: I made the change you suggested. Thanks for catching that!

**Pull request merged**